compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -2515,21 +2515,22 @@ impl<'tcx> TyCtxt<'tcx> {
         (ident, scope)
     }
 
-    /// Returns `true` if the debuginfo for `span` should be collapsed to the outermost expansion
-    /// site. Only applies when `Span` is the result of macro expansion.
-    ///
-    /// - If the `collapse_debuginfo` feature is enabled then debuginfo is not collapsed by default
-    ///   and only when a macro definition is annotated with `#[collapse_debuginfo]`.
-    /// - If `collapse_debuginfo` is not enabled, then debuginfo is collapsed by default.
-    ///
-    /// When `-Zdebug-macros` is provided then debuginfo will never be collapsed.
-    pub fn should_collapse_debuginfo(self, span: Span) -> bool {
-        !self.sess.opts.unstable_opts.debug_macros
-            && if self.features().collapse_debuginfo {
-                span.in_macro_expansion_with_collapse_debuginfo()
-            } else {
-                span.from_expansion()
-            }
+    /// Macros in standart library (core/alloc/std)
+    pub fn in_std_macro(self, macro_def_id: DefId) -> bool {
+        let crate_name = self.crate_name(macro_def_id.krate);
+        [sym::core, sym::alloc, sym::std].contains(&crate_name)
+    }
+
+    pub fn collapsed_debuginfo(self, span: Span, upto: SyntaxContext) -> Span {
+        let collapse_enabled = !self.sess.opts.unstable_opts.debug_macros;
+        let attr_enabled = self.features().collapse_debuginfo;
```

Comment:
Ah, I see why there were several seemingly identical tests, it's due to this code.

This is a misuse of the `#![feature]` attribute, this attribute shouldn't change any behavior, only enable or disable feature gate errors.
If you want a switch to enable/disable `collapse_debuginfo` for built-in/standard macros in tests, then it's better to add a new crate-level `#[rustc_*]` attribute.
Or even better not rely on libstd macros in tests at all, and instead add the test's own `collapse_debuginfo` and non-`collapse_debuginfo` macros.